structuredheadings/plugin.js

```diff
@@ -649,7 +602,66 @@
           }
         }
       },
+      applyHeadingPreset: {
+        clearAutonumberClassesForHeading: function (editor, heading) {
+          clearStyles(editor, heading);
+          cssUtils.clearLevel(editor, heading);
+          cssUtils.clearNumbering(editor, heading);
+        },
 
+        setAutonumberClassesForHeading: function (editor, value, heading) {
+          cssUtils.setNumbering(editor, heading);
+          cssUtils.setLevel(editor, heading);
+          setCurrentStyle(editor, heading, value);
+        },
+
+        handleCollapsedSelection: function (editor) {
+          // avoid an extra snapshot created by the matchHeading command
+          editor.fire("lockSnapshot", { dontUpdate: true });
+          editor.execCommand("matchHeading");
+          editor.fire("unlockSnapshot");
+
+          // return the one matched current heading
+
+          return [CKEDITOR.plugins.structuredheadings.getCurrentBlockFromPath(editor)];
+        },
+
+        exec: function (editor, value) {
+          var selection = editor.getSelection();
+          var headings = CKEDITOR.plugins.structuredheadings.getHeadingsInSelection(
+                editor,
+                selection
+              );
+          var func;
+
+          if (value === "clear") {
+            func = this.clearAutonumberClassesForHeading.bind(this, editor);
+          } else {
+            func = this.setAutonumberClassesForHeading.bind(this, editor, value);
+
+             // prep the html for the collapsed, not in a heading case
+            if (headings === null) {
+              // put the new heading into the headings array
+              headings = this.handleCollapsedSelection(editor);
+            }
+          }
+
+          if (headings) {
+            for (var i = 0; i < headings.length; i++) {
+              func(headings[i]);
+            }
+          }
+
+          editor.fire("lockSnapshot");
+          // the snapshot needs to be locked here, because
+          // execCommand will also create a snapshot, leading to
+          // an intermediate snapshot with some of the styles applied, but not all
+
+          // apply the correct bulletstyle for all numbered headings
+          editor.execCommand("reapplyStyle", value);
+          editor.fire("unlockSnapshot");
```

Comment:
this snapshotting is still a headache, i tried to clear it up but kept breaking undo cases.  argh.